using `map[string]!string` should be a checker error, not a cgen one

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.3.4 2e9c469.3e3b289
OS: linux, Ubuntu 20.04.5 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz

getwd: /v/vnew
vexe: /v/vnew/v
vexe mtime: 2023-06-24 11:41:24

vroot: OK, value: /v/vnew
VMODULES: OK, value: /home/delian/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.40.0
Git vroot status: weekly.2023.25-27-g3e3b2895
.git/config present: true

CC version: cc (Ubuntu 10.3.0-1ubuntu1~20.04) 10.3.0
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg x.v`
```v
dump(map[string]!string{})

```

**What did you expect to see?**

A checker error, telling me that `map[string]!string` is not supported (since result types are only valid as return types of functions).


**What did you see instead?**
```
==================
/tmp/v_1000/x.1535538020173294804.tmp.c:2127: error: cannot convert 'struct _result_string' to 'unsigned char *'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/containers.v
+++ b/vlib/v/checker/containers.v
@@ -349,6 +349,9 @@ fn (mut c Checker) map_init(mut node ast.MapInit) ast.Type {
 	if node.typ != 0 {
 		info := c.table.sym(node.typ).map_info()
 		if info.value_type != 0 {
+			if info.value_type.has_flag(.result) {
+				c.error('cannot use Result type as map value type', node.pos)
+			}
 			val_sym := c.table.sym(info.value_type)
 			if val_sym.kind == .struct_ {
 				val_info := val_sym.info as ast.Struct

--- a/vlib/v/checker/struct.v
+++ b/vlib/v/checker/struct.v
@@ -129,6 +129,12 @@ fn (mut c Checker) struct_decl(mut node ast.StructDecl) {
 			if sym.kind == .none_ {
 				c.error('cannot use `none` as field type', field.type_pos)
 			}
+			if sym.kind == .map {
+				info := sym.map_info()
+				if info.value_type.has_flag(.result) {
+					c.error('cannot use Result type as map value type', field.type_pos)
+				}
+			}
 
 			if field.has_default_expr {
 				c.expected_type = field.typ

--- a/vlib/v/checker/tests/map_with_result_value_err.out
+++ b/vlib/v/checker/tests/map_with_result_value_err.out
@@ -1,0 +1,5 @@
+vlib/v/checker/tests/map_with_result_value_err.vv:2:8: error: cannot use Result type as map value type
+    1 | fn main() {
+    2 |   dump(map[string]!string{})
+      |        ~~~~~~~~~~~~~~~~~~~~
+    3 | }

--- a/vlib/v/checker/tests/map_with_result_value_err.out
+++ b/vlib/v/checker/tests/map_with_result_value_err.out
@@ -1,5 +1,19 @@
-vlib/v/checker/tests/map_with_result_value_err.vv:2:8: error: cannot use Result type as map value type
-    1 | fn main() {
-    2 |   dump(map[string]!string{})
-      |        ~~~~~~~~~~~~~~~~~~~~
+vlib/v/checker/tests/map_with_result_value_err.vv:2:7: error: cannot use Result type as map value type
+    1 | struct Foo {
+    2 |     map1 map[string]!string
+      |          ~~~~~~~~~~~~~~~~~~
     3 | }
+    4 |
+vlib/v/checker/tests/map_with_result_value_err.vv:6:8: error: cannot use Result type as map value type
+    4 | 
+    5 | fn main() {
+    6 |   _ := map[string]!string{}
+      |        ~~~~~~~~~~~~~~~~~~~~
+    7 | 
+    8 |   dump(map[string]!string{})
+vlib/v/checker/tests/map_with_result_value_err.vv:8:8: error: cannot use Result type as map value type
+    6 |   _ := map[string]!string{}
+    7 | 
+    8 |   dump(map[string]!string{})
+      |        ~~~~~~~~~~~~~~~~~~~~
+    9 | }

--- a/vlib/v/checker/tests/map_with_result_value_err.vv
+++ b/vlib/v/checker/tests/map_with_result_value_err.vv
@@ -1,3 +1,9 @@
+struct Foo {
+	map1 map[string]!string
+}
+
 fn main() {
+  _ := map[string]!string{}
+
   dump(map[string]!string{})
 }

--- a/vlib/v/checker/tests/map_with_result_value_err.vv
+++ b/vlib/v/checker/tests/map_with_result_value_err.vv
@@ -1,0 +1,3 @@
+fn main() {
+  dump(map[string]!string{})
+}

--- a/vlib/v/tests/map_value_with_option_result_test.v
+++ b/vlib/v/tests/map_value_with_option_result_test.v
@@ -15,13 +15,11 @@ fn foo(arg map[string]?string) ?string {
 }
 
 struct Foo {
-	map1 map[string]!string
 	map2 map[string]?string
 }
 
 fn bar() {
 	map1 := map[string]?string{}
-	map2 := map[string]!string{}
 }
 
 fn baz(arg map[string]?string) ?string {

